src/basic/fs-util.c

```diff
@@ -156,24 +156,16 @@ int readlink_malloc(const char *p, char **ret) {
 
 int readlink_value(const char *p, char **ret) {
         _cleanup_free_ char *link = NULL;
-        char *value;
         int r;
 
+        assert(p);
+        assert(ret);
+
         r = readlink_malloc(p, &link);
         if (r < 0)
                 return r;
 
-        value = basename(link);
-        if (!value)
-                return -ENOENT;
-
-        value = strdup(value);
-        if (!value)
-                return -ENOMEM;
-
-        *ret = value;
-
-        return 0;
+        return path_extract_filename(link, ret);
```

Comment:
so, path_extract_filename() will return O_DIRECTORY if the path ends in a `/`. Which we shouldn't allow here I guess. Maybe hence add:

```c
_cleanup_free char *value = NULL;

r = path_extract_filename(link, &value);
if (r < 0)
        return r;
if (r == O_DIRECTORY)
        return _EINVAL;
*ret = TAKE_PTR(value);
return 0;
```

Reply:
Ouch!